editor/components/block-list/index.js

```diff
@@ -3,60 +3,66 @@
  */
 import { connect } from 'react-redux';
 import {
-	filter,
+	reduce,
 	get,
 	map,
 } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { createElement } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
 import './style.scss';
 import BlockListLayout from './layout';
-import { getBlocks } from '../../store/selectors';
+import { getBlocks, getBlockOrder } from '../../store/selectors';
 
-function BlockList( {
+const UngroupedLayoutBlockList = connect(
+	( state, ownProps ) => ( {
+		blockUIDs: getBlockOrder( state, ownProps.rootUID ),
+	} )
+)( BlockListLayout );
+
+const GroupedLayoutBlockList = connect(
+	( state, ownProps ) => ( {
+		blocks: getBlocks( state, ownProps.rootUID ),
+	} ),
+)( ( {
 	blocks,
-	renderBlockMenu,
-	layouts = {},
-	rootUID,
-	showContextualToolbar,
-} ) {
+	layouts,
+	...props
+} ) => map( layouts, ( layout ) => {
+	// Filter blocks assigned to layout when rendering grouped layouts.
+	const layoutBlockUIDs = reduce( blocks, ( result, block ) => {
+		if ( get( block, [ 'attributes', 'layout' ] ) === layout.name ) {
+			result.push( block.uid );
+		}
+
+		return result;
+	}, [] );
+
+	return (
+		<BlockListLayout
+			key={ layout.name }
+			layout={ layout.name }
+			isGroupedByLayout
+			blockUIDs={ layoutBlockUIDs }
+			{ ...props }
+		/>
+	);
+} ) );
+
+const BlockList = ( props ) => createElement(
```

Comment:
We might want to introduce [branch (recompose)](https://github.com/acdlite/recompose/blob/master/docs/API.md#branch) HOC if we have more use cases.